ARCHITECTURE.md

```diff
@@ -0,0 +1,275 @@
+# HyperDB Architecture
+
+HyperDB is a scalable peer-to-peer key-value database.
+
+## Filesystem metaphor
+
+HyperDB is structured to be used much like a traditional hierarchical
+filesystem. A value can be written and read at locations like `/foo/bar/baz`,
+and the API supports querying or tracking values at subpaths, like how watching
+for changes on `/foo/bar` will report both changes to `/foo/bar/baz` and also
+`/foo/bar/19`.
+
+## Set of append-only logs
+
+A HyperDB is fundamentally a set of
+[hypercore](https://github.com/mafintosh/hypercore)s. A *hypercore* is a secure
+append-only log that is identified by a public key, and can only be written to
+by the holder of the corresponding private key. Because it is append-only, old
+values cannot be deleted nor modified. Because it is secure, any other peer in
+the network can verify these attributes without inherently trusting the author.
+
+Each entry in a hypercore has a *sequence number*, that increments by 1 with
+each write, starting at 0 (`seq=0`).
+
+HyperDB builds its hierarchical key-value store on top of these hypercores, and
+also provides facilities for authorization, and replication of those member
+hypercores.
+
+### Directed acyclic graph
+
+The combination of all operations performed on a HyperDB by all of its members
+forms a DAG (*directed acyclic graph*). Each database operations link backward
+to all of the known "heads" in the graph.
+
+For example, let's say Alice starts a new HyperDB and writes 2 values to it:
+
+```
+// Feed
+
+0 (/foo/bar = 'baz')
+1 (/foo/2   = '{ "some": "json" }')
+
+
+// Graph
+
+Alice:  0  <---  1
+```
+
+Where sequence number 1 (the second entry) refers to sequence number 0 on the
+same feed (Alice's).
+
+Now Alice *authorizes* Bob to write to the HyperDB. Internally, this means Alice
+writes a special message to her feed saying that Bob's feed (identified by his
```

Comment:
why are we talking about feeds now? is it a synonym for log? also, what makes any feed Alice's feed?

Reply:
Yes, a feed is an append-only log. The terminology can get confusing. You can also roughly interchange the term "hypercore" with either in the context of hyperdb.

A feed is Alice's because she signs the feed entries with her private key. Nothing links this key to her at a biological level. It's only "hers" in this sense so long as she keeps that key safe on her computer and doesn't lose it or share it with somebody else.